### PR TITLE
[WIP] Update to latest Rust (nightlies).

### PR DESCRIPTION
**Pull request by toqueteos**
**MERGE WHENEVER YOU WANT.**

I'm working my way up to getting hematite working on Rust 1.0 ~~alpha~~ nightlies.

If anyone wants to suggest something I welcome each and everyone of you.

### CODE REVIEW

**Comment by eddyb** [comment]
A lot of this is long-awaited cleanup, thanks!

**Comment by toqueteos** [comment]
@eddyb Mind helping a bit with the Cell related issues? Check travis log. Pretty please.

File: src/minecraft/nbt.rs [hunk 1]
```diff
@@ -44,8 +46,26 @@ impl fmt::Show for Nbt {
     }
 }
 
+impl ToString for Nbt {
+    fn to_string(&self) -> String {
+        match *self {
```
Comment (eddyb) [inline src/minecraft/nbt.rs]: DRY - you have 3 options here:
1. implement `ToString` with `format!("{:?}", *self)`
2. replace `nbt.to_string()` calls with `format!("{:?}", nbt)`
3. add a `fmt::String` impl fot `Nbt` that is similar to the `fmt::Show` impl, but doesn't add all the type suffixes and whatnot
1 and 2 preserve the old behavior, but 1 is unidiomatic. If there's only a few uses, 2 would be the best choice.

**Comment by eddyb** [comment]
The `Cell` issues seem to stem from missing `#[derive(Copy)]` on a few types, like `shader::Buffer`.
I wish the new trait logic in rustc would pick potentially matching implementations and list the missing bounds, because the fix would be obvious from the error message.

**Comment by toqueteos** [comment]
@eddyb Yesterday people suggested me using a RefCell instead of Cell if missing Copy is my issue. It adds complexity but it's one way to deal with it. Is Cell really required?

As for the to_string issues that's what I'm working on right now.

**Comment by toqueteos** [comment]
Down to 10 errors! I can't manage to find a way to solve these.

**Comment by eddyb** [comment]
Yes! `Cell` is really needed here, the type is copyable, just didn't get `#[derive(Copy)]` when it started being required.

**Comment by toqueteos** [comment]
@eddyb That's easy to fix!

**Comment by eddyb** [comment]
Sounds like we're missing some gfx-rs annotation or something? Sorry, I can't investigate more, schoolwork has caught up to me :(.

**Comment by bvssvni** [comment]
Try changing the Buffer to:

```Rust
#[derive(Copy)]
pub struct Buffer {
    buf: gfx::BufferHandle<Vertex>,
    batch: Program
}
```

**Comment by toqueteos** [comment]
@bvssvni 

```
   Compiling hematite v0.0.0 (file:///home/ratnu/code/rust/github.com/toqueteos/hematite)
src/shader.rs:112:10: 112:14 error: the trait `Copy` may not be implemented for this type; field `batch` does not implement `Copy`
src/shader.rs:112 #[derive(Copy)]
                           ^~~~
note: in expansion of #[derive]
src/shader.rs:112:1: 112:16 note: expansion site
error: aborting due to previous error
Could not compile `hematite`.
```

I tried adding `#[derive(Copy)]` to `ShaderParam` (the type annotated as Program) but same error.

**Comment by bvssvni** [comment]
I've sent an PR to derive `Copy` on `RefBatch` https://github.com/gfx-rs/gfx-rs/pull/502

**Comment by bvssvni** [comment]
@toqueteos You can override dependencies locally with this trick https://github.com/PistonDevelopers/piston/issues/482 so you don't have to wait for PRs to merge.

**Comment by toqueteos** [comment]
@bvssvni D'oh. Right! Let me try that and then I'm off to bed, it's getting late :/

**Comment by toqueteos** [comment]
IT WORKS! Down to THREE errors.

```
   Compiling hematite v0.0.0 (file:///home/ratnu/code/rust/github.com/toqueteos/hematite)
src/minecraft/model.rs:238:33: 238:51 error: mismatched types:
 expected `&mut (minecraft::model::Face, collections::string::String)`,
    found `&_`
(values differ in mutability)
src/minecraft/model.rs:238                             for &(ref mut face, _) in model.faces.slice_from_mut(element_start).iter_mut() {
                                                           ^~~~~~~~~~~~~~~~~~
src/minecraft/model.rs:239:33: 239:45 error: the type of this value must be known in this context
src/minecraft/model.rs:239                                 face.ao_face = None;
                                                           ^~~~~~~~~~~~
src/minecraft/model.rs:242:66: 242:66 help: methods from traits can only be called if the trait is implemented and in scope; the following traits define a method `iter_mut`, perhaps you need to implement one of them:
src/minecraft/model.rs:242:66: 242:66 help: candidate #1: `core::slice::SliceExt`
src/minecraft/model.rs:242:66: 242:66 help: candidate #2: `core::slice::SliceExt`
src/minecraft/model.rs:242:66: 242:66 help: candidate #3: `collections::slice::SliceExt`
src/minecraft/model.rs:242:66: 242:66 help: candidate #4: `collections::slice::SliceExt`
src/minecraft/model.rs:242:66: 242:66 help: candidate #5: `collections::slice::SliceExt`
src/minecraft/model.rs:249:70: 249:70 help: methods from traits can only be called if the trait is implemented and in scope; the following traits define a method `iter_mut`, perhaps you need to implement one of them:
src/minecraft/model.rs:249:70: 249:70 help: candidate #1: `core::slice::SliceExt`
src/minecraft/model.rs:249:70: 249:70 help: candidate #2: `core::slice::SliceExt`
src/minecraft/model.rs:249:70: 249:70 help: candidate #3: `collections::slice::SliceExt`
src/minecraft/model.rs:249:70: 249:70 help: candidate #4: `collections::slice::SliceExt`
src/minecraft/model.rs:249:70: 249:70 help: candidate #5: `collections::slice::SliceExt`
src/minecraft/model.rs:255:66: 255:66 help: methods from traits can only be called if the trait is implemented and in scope; the following traits define a method `iter_mut`, perhaps you need to implement one of them:
src/minecraft/model.rs:255:66: 255:66 help: candidate #1: `core::slice::SliceExt`
src/minecraft/model.rs:255:66: 255:66 help: candidate #2: `core::slice::SliceExt`
src/minecraft/model.rs:255:66: 255:66 help: candidate #3: `collections::slice::SliceExt`
src/minecraft/model.rs:255:66: 255:66 help: candidate #4: `collections::slice::SliceExt`
src/minecraft/model.rs:255:66: 255:66 help: candidate #5: `collections::slice::SliceExt`
src/minecraft/nbt.rs:293:5: 295:6 error: method `decode` has 0 type parameters but its trait declaration has 1 type parameter [E0049]
src/minecraft/nbt.rs:293     fn decode(d: &mut Decoder) -> DecodeResult<Nbt> {
src/minecraft/nbt.rs:294         d.pop()
src/minecraft/nbt.rs:295     }
error: aborting due to 3 previous errors
```

**Comment by bvssvni** [comment]
To fix this temporarily, replace the closure with `|x: &json::Json| -> Option<json::Array> x.as_array().map(|x| x.clone())`.

**Comment by toqueteos** [comment]
@bvssvni Indeed, it solves the problem. Now a new load of errors beholds...

File: src/minecraft/block_state.rs [hunk 2]
```diff
@@ -422,25 +419,25 @@ pub fn fill_buffer(block_states: &BlockStates, biomes: &Biomes, buffer: &mut Vec
                         loop {
                             let (cond, idx) = match model.polymorph_oracle[i] {
```
Comment (toqueteos) [inline src/minecraft/block_state.rs]:
```rust
let (cond, idx) = match model.polymorph_oracle[i].clone()
```

**Comment by toqueteos** [comment]
Also `block_state.rs:457` (can't comment on that line):

```rust
let block_xyz = match model.random_offset.clone() {
```

File: src/minecraft/nbt.rs [hunk 3]
```diff
@@ -411,7 +439,9 @@ impl serialize::Decoder<DecoderError> for Decoder {
         Ok(value)
     }
 
-    fn read_tuple<T>(&mut self, tuple_len: uint, f: |&mut Decoder| -> DecodeResult<T>) -> DecodeResult<T> {
+    fn read_tuple<T, F>(&mut self, tuple_len: usize, f: F) -> DecodeResult<T> 
+        where F: FnOnce(&mut Self) -> DecodeResult<T> 
+    {
         self.read_seq(|d, len| {
```
Comment (toqueteos) [inline src/minecraft/nbt.rs]:
```rust
self.read_seq(move |d, len| {
```

**Comment by toqueteos** [comment]
It seems the monomorphization error is here to stay until I solve the recursive mess. PartialModel::load is called just on Model::load and inside it when models have a parent attribute. It shouldn't be infinite, maybe long in some cases but not infinite...

**Comment by bvssvni** [comment]
Is there a way to get more information about what causes the monomorphization error?

**Comment by toqueteos** [comment]
@bvssvni Nothing more. That's all log. PartialModel::load is calling itself, taht's where I'm digging now.

**Comment by bvssvni** [comment]
The way I solved this in ai_behavior is by using `f: &mut F` instead of `mut f: F`.

**Comment by toqueteos** [comment]
@bvssvni Let's see how that works!

**Comment by toqueteos** [comment]
I've tinkered a bit and isolated the bug location, it's due to the recursive call on `PartialModel::load` [LINK](https://github.com/toqueteos/hematite/blob/d8cd6c388149744faf563ba702180a00a7006340/src/minecraft/model.rs#L116).

I've also tried removing that block (same action as None for Some) and hematite finished building, so we are close guys. We need to refactor a bit `model.rs`.

**Comment by toqueteos** [comment]
It's Christmas guys, IT'S WORKING.

**Comment by toqueteos** [comment]
Explanation of the monomorphization error, what I understood:

Our `PartialModel::load` called recursively itself when models had a parent defined:

```rust
let mut model = match obj.find("parent").and_then(|x| x.as_string()) {
    Some(parent) => PartialModel::load(parent, assets, atlas, cache, |m, _| m.clone())
    // ...
```

You see that closure on the right? `|m, _| m.clone()` That thing was *translated* to a `FnMut(&PartialModel, &mut AtlasBuilder) -> T` but we just need `T` to be a `PartialModel`.

So I just defined a new fn outside:

```rust
fn clone_parent(m: &PartialModel, a: &mut AtlasBuilder) -> PartialModel {
    m.clone()
}
```

and voilá, problem solved.

**Comment by bvssvni** [comment]
:balloon:

**Comment by bvssvni** [comment]
Can we merge now?

**Comment by toqueteos** [comment]
@bvssvni Yeah, whenever you feel right about the mess I did. I was waiting so eddy or anyone could review whatever they want. Don't wanna rush anything. Feel free to merge whenever you want.

**Comment by eddyb** [comment]
Hmm, using a function is not the proper solution for this. Let me look at it some more.

File: src/minecraft/model.rs [hunk 4]
```diff
@@ -91,28 +92,40 @@ impl Opacity {
...
+    m.clone()
+}
+
 impl PartialModel {
-    fn load<T>(name: &str, assets: &Path, atlas: &mut AtlasBuilder,
+    fn load<T, F>(name: &str, assets: &Path, atlas: &mut AtlasBuilder,
                cache: &mut HashMap<String, PartialModel>,
-               f: |&PartialModel, &mut AtlasBuilder| -> T) -> T {
+               mut f: F) -> T 
+        where F: FnMut(&PartialModel, &mut AtlasBuilder) -> T
```
Comment (eddyb) [inline src/minecraft/model.rs]: You could have `f: &mut FnMut(&PartialModel, &mut AtlasBuilder) -> T`.
What's interesting here is that it seems to me like it's a bug, because the `|_, m| m.clone()` closure doesn't capture or work with anything mentioning `T` or `F`. Though I don't think it's easy to fix in the compiler.

**Comment by bvssvni** [comment]
When trying to opening a world I get this:

```
thread '<main>' panicked at 'index out of bounds: the len is 0 but the index is 16', src/main.rs:1
```

File: src/minecraft/model.rs (hunk 4 above)
Comment (toqueteos) [inline src/minecraft/model.rs]: Changing `mut f: F` to `f: &mut F` gives the same exact error. Moving the closure to a external typed function seems fine. What's wrong with it?

**Comment by toqueteos** [comment]
@bvssvni I get:

```
HashMap {"Data": HashMap {"version": 19133i, "Difficulty": 2b, ...
...
BIG ASS HASHMAP LOG
...
rainTime": 170718i, "BorderCenterZ": 0.0, "allowCommands": 0b}}
libGL error: failed to authenticate magic 3
libGL error: failed to load driver: vboxvideo
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: "Could not create GL context"', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/result.rs:742
```

**Comment by bvssvni** [comment]
@toqueteos VBox doesn't support OpenGL 3.2. Try 2.1 or something.

**Comment by toqueteos** [comment]
@bvssvni Well I reverted the libc changes on `region.rs` so it now also builds on Windows but complains about `CreateFileMappingW failure = 6` (mmap equivalent on Win-land). Gotta try with regular file open.

**Comment by toqueteos** [comment]
Isolated problem.

Inside `BlockStates::load_with_states` we defined a Vec called models:

```rust
let mut models = Vec::with_capacity(last_id as usize + 1);
```
[Source](https://github.com/toqueteos/hematite/blob/772845bb2be3f37b6aa460580a75f62a3867bf2a/src/minecraft/block_state.rs#L209)

On this Vec we store the processed json info from MC's assets folder.

---

The indexing error is happening here:

```rust
models[state.id as usize] = ModelAndBehavior {
    model: model,
    random_offset: state.random_offset,
    polymorph_oracle: state.polymorph_oracle
};
```
[Source](https://github.com/toqueteos/hematite/blob/772845bb2be3f37b6aa460580a75f62a3867bf2a/src/minecraft/block_state.rs#L347)

---

It is weird because we are calling Vec::reserve right before doing this assignment... Maybe we could use a HashMap instead of a Vec. Seems like a disperse array.

File: src/minecraft/model.rs (hunk 4 above)
Comment (eddyb) [inline src/minecraft/model.rs]: I said `&mut FnMut(&PartialModel, &mut AtlasBuilder) -> T` which is quite different.
If you want to keep the free function, could it be a generic version of `|_, x| x.clone(0)`?
```rust
fn clone_2nd<T, U: Clone>(_: T, x: &U) -> U { x.clone() }
```

**Comment by eddyb** [comment]
I don't like the `HashMap` in the output - NBT should do its own formatting, maybe `Compound` or no prefix at all.

File: src/minecraft/block_state.rs [hunk 5]
```diff
@@ -344,10 +340,11 @@ impl BlockStates {
             rotate_faces(&mut model, 0, 2, variant.rotate_y);
 
             let len = models.len();
-            if state.id as uint >= len {
-                models.grow(state.id as uint - len + 1, ModelAndBehavior::empty());
+            if state.id as usize >= len {
+                models.reserve(state.id as usize - len + 1);
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: This is not doing the same thing as before, `reserve` deals with the capacity, I was adding elements onto the vector.
And no, you can't use a `HashMap`, that would ruin the indexing behavior, in a pretty hot loop.

Comment (toqueteos) [inline src/minecraft/block_state.rs]: Then it should have been a resize right? There's no grow now.

Comment (eddyb) [inline src/minecraft/block_state.rs]: Oh, they added `resize`, that's neat. Yeah, `models.resize(state.id as usize)` should work.

**Comment by toqueteos** [comment]
Guys is this enough to merge?
hematite screenshot on Win 8.1

**Comment by toqueteos** [comment]
Damn, I deleted the AsRawFd import... One second.

File: src/minecraft/region.rs [hunk 6]
```diff
@@ -60,20 +61,33 @@ impl Region {
         unsafe { mem::transmute(slice) }
     }
 
+    #[cfg(windows)]
+    pub fn open(filename: &Path) -> IoResult<Region> {
+        let mut file = try!(File::open(filename));
+        let res = Region {
+            mmap: try!(file.read_to_end()),
```
Comment (eddyb) [inline src/minecraft/region.rs]: Wait, what? Surely there is a way on windows? I didn't realize you were doing this.

Comment (toqueteos) [inline src/minecraft/region.rs]: I tried going the "old" libc way but MemoryMap complained a lot so I made a workaround... It's UGLY AS FRIKEN BOLLOCKS but works...

**Comment by toqueteos** [comment]
I just pushed a new commit fixing the stupid renaming error so I'll copy my response to @eddyb here:

> Wait, what? Surely there is a way on windows? I didn't realize you were doing this.

I tried going the "old" libc way but os::MemoryMap complained a lot (the Fd on windows is a libc::HANDLE instead of a c_int, and libc::open returns some other weird type libc::something::c95::something) so I made a workaround... It's UGLY AS FRIKEN BOLLOCKS but works...

**Comment by eddyb** [comment]
[`libc::open`](http://doc.rust-lang.org/libc/libc/funcs/posix88/fcntl/fn.open.html) returns `libc::c_int`, the function which actually returns `libc::HANDLE`, which is an winapi type, is, in fact, a winapi function, [`CreateFileW`](http://doc.rust-lang.org/libc/src/libc/lib.rs.html#4971-4977).

**Comment by toqueteos** [comment]
@eddyb That's the right function, but libc doesn't provide `GetFileSizeEx`. I'm looking for any winapi crate with that function but no one defines it, yet. Would it be ok if we opened the file first with regular `File::open`, then call `::stat` and then using `CreateFileW`?

**Comment by eddyb** [comment]
How does `std::sys` implement it? Could we duplicate that until it's exposed like the fd is on any unix?

**Comment by toqueteos** [comment]
@eddyb Problem solved, sorry about all the edits. I've just updated the commit. It does run successfully on Win, it should too on Linux.

**Comment by eddyb** [comment]
It seems that some libs need to get updated (the `ViewItem` thing is my own personal fault, sorry!).

**Comment by toqueteos** [comment]
@eddyb It also seems slice syntax finally made it to nightlies... Loads of changes to do, again.

**Comment by eddyb** [comment]
One more thing I just noticed: you shouldn't use `Box::new`, you risk blowing the stack (and it's inefficient`. `box expr` is only feature-gated because it will be somewhat redesigned before 1.0.

**Comment by toqueteos** [comment]
@eddyb It was Box before, I just updated to new syntax. No Box instead or Cell?

**Comment by eddyb** [comment]
`Box::new` isn't a new syntax, it's a workaround, almost everyone who is using feature gates should just continue using `box expr`.

File: src/minecraft/biome.rs [hunk 7]
```diff
@@ -1,24 +1,25 @@
...
 
 pub struct Biomes {
-    biomes: Box<[Option<Biome>, ..256]>
+    biomes: Box<[Option<Biome>; 256]>
 }
 
 impl Biomes {
     pub fn load(assets: &Path) -> Biomes {
-        let mut biomes = box() ([None, ..256]);
+        let mut biomes = Box::new([None; 256]);
```
Comment (eddyb) [inline src/minecraft/biome.rs]: This, specifically, can be `box [None; 256]` - I think the `()` was from when `~[a, b, c]` produced `~[T]` while `~([a, b, c])` would have been `~([T, ..3])` and the former had a different, incompatible, layout in memory.

Comment (toqueteos) [inline src/minecraft/biome.rs]: They seem quite the same thing... http://is.gd/QwepDe

Comment (eddyb) [inline src/minecraft/biome.rs]: Not sure what are you pointing out.
[`Box::new(x)` can overflow the stack](http://is.gd/IOXe6H) while [`box x` doesn't](http://is.gd/DSHUbt).

Comment (toqueteos) [inline src/minecraft/biome.rs]: Didn't know that, I just asked on IRC. Will fix tomorrow it's 3:45 AM.

**Comment by toqueteos** [comment]
Updating to latest nightlies, gfx-rs is holding us back because `fmt::{ Show, String }` got renamed and now `Result.unwrap` doesn't work with some errors not implementing `fmt::Display` (String).

**Comment by toqueteos** [comment]
Ready to merge. r? @eddyb @bvssvni

File: src/minecraft/block_state.rs [hunk 8]
```diff
@@ -279,16 +273,18 @@ impl BlockStates {
...
                                         &mut atlas, &mut partial_model_cache);
 
-            let rotate_faces = |m: &mut Model, ix, iy, rot_mat: [i32, ..4]| {
+            let rotate_faces = |&: m: &mut Model, ix: usize, iy: usize, rot_mat: [i32; 4]| {
                 let [a, b, c, d] = rot_mat.map(|x: i32| x as f32);
                 for face in m.faces.iter_mut() {
                     for vertex in face.vertices.iter_mut() {
-                        let xyz = &mut vertex.xyz;
-                        let [x, y] = [ix, iy].map(|i| xyz[i] - 0.5);
+                        let mut xyz: [f32; 3] = vertex.xyz;
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: This is... very wrong. Good thing I took another look at it - you're copying out the value, while the original code was mutating it in-place (`xyz` needs to point to `vertex.xyz`).

File: src/minecraft/block_state.rs [hunk 9]
```diff
@@ -328,15 +324,13 @@ impl BlockStates {
...
+
+            let rotate_faces = |&: m: &mut Model, ix: usize, iy: usize, r: OrthoRotation| {
                 match r {
                     Rotate0 => {}
-                    Rotate90 => rotate_faces(m, ix, iy, [0,-1,
-                                                         1, 0]),
-                    Rotate180 => rotate_faces(m, ix, iy, [-1, 0,
-                                                           0,-1]),
-                    Rotate270 => rotate_faces(m, ix, iy, [0, 1,
-                                                         -1, 0]),
```
Comment (eddyb) [inline src/minecraft/block_state.rs]: These looked like they did because they're 2D rotation matrices.

File: src/minecraft/model.rs [hunk 10]
```diff
@@ -91,28 +91,40 @@ impl Opacity {
...
+        let obj = json::Json::from_reader(&mut File::open(&path).unwrap()).unwrap();
 
-        let mut model = match json.find_with(|k| "parent".cmp(k.as_slice())).and_then(|x| x.as_string()) {
-            Some(parent) => PartialModel::load(parent, assets, atlas, cache, |m, _| m.clone()),
+        let mut model = match obj.find("parent").and_then(|x| x.as_string()) {
+            // FIXME(toqueteos): Cthulu himself came here and inspired me, if we use a closure here instead of
+            // "clone_parent" this would trigger an error: "reached the recursion limit during monomorphization"
+            Some(parent) => PartialModel::load(parent, assets, atlas, cache, clone_parent),
             None => PartialModel {
+            // _ => PartialModel {
```
Comment (eddyb) [inline src/minecraft/model.rs]: Why is this line here?

File: src/minecraft/nbt.rs [hunk 11]
```diff
@@ -247,27 +272,26 @@ impl Decoder {
...
     }
 }
 
-impl Decodable<Decoder, DecoderError> for Nbt {
-    fn decode(d: &mut Decoder) -> DecodeResult<Nbt> {
-        d.pop()
-    }
-}
+// impl Decodable for Nbt {
+//     fn decode<D: serialize::Decoder>(d: &mut D) -> Result<Self, serialize::Decoder::Error> {
```
Comment (eddyb) [inline src/minecraft/nbt.rs]: Uhh, did they really require `Decodable` to be generic over the `Decoder` in the method?

File: src/minecraft/nbt.rs [hunk 12]
```diff
@@ -316,7 +344,7 @@ impl serialize::Decoder<DecoderError> for Decoder {
     fn read_f32(&mut self) -> DecodeResult<f32> { expect!(self, Float) }
 
     fn read_char(&mut self) -> DecodeResult<char> {
-        let s = try!(self.read_str());
+        let s : String = try!(self.read_str());
```
Comment (eddyb) [inline src/minecraft/nbt.rs]: Is this required?

File: src/minecraft/region.rs [hunk 13]
```diff
@@ -15,38 +19,48 @@ use chunk::{
...
-            libc::close(fd);
-            Ok(res)
-        }
+    #[cfg(windows)]
+    pub fn open(filename: &Path) -> IoResult<Region> {
+        let file = try!(File::open(filename));
+        let stat: FileStat = try!(file.stat());
+        let min_len = stat.size as usize;
+        let options = &[
+            os::MapOption::MapFd(file.as_raw_handle()),
```
Comment (eddyb) [inline src/minecraft/region.rs]: Maybe move creating `MapOption::MapFd` from `File` to a function?

File: src/shader.rs [hunk 14]
```diff
@@ -167,7 +170,7 @@ impl<D: Device<C>, C: CommandBuffer> Renderer<D, C> {
                     &mesh,
                     mesh.to_slice(gfx::PrimitiveType::TriangleList),
                     &self.drawstate
-                ).unwrap()
+                ).ok().unwrap()
```
Comment (eddyb) [inline src/shader.rs]: :(

Comment (bvssvni) [inline src/shader.rs]: Perhaps `.expect("Error message")` would be better here?

Comment (toqueteos) [inline src/shader.rs]: This is just temporal, I opened an issue explaining why on gfx_texture: https://github.com/PistonDevelopers/gfx_texture/issues/20

It's due to Result requiring errors to implement fmt::Display.

File: src/minecraft/region.rs (hunk 13 above)
Comment (toqueteos) [inline src/minecraft/region.rs]: Yes. I'm on it. Seems nicer and also less *"os dependent lines"*.

File: src/minecraft/model.rs (hunk 10 above)
Comment (toqueteos) [inline src/minecraft/model.rs]: Leftovers.

File: src/minecraft/block_state.rs (hunk 8 above)
Comment (toqueteos) [inline src/minecraft/block_state.rs]: If I recall correctly I changed this because Rust couldn't infer the type of [ix, iy] and was complaining a lot. Let's see how it does now.

File: src/minecraft/nbt.rs (hunk 12 above)
Comment (toqueteos) [inline src/minecraft/nbt.rs]: 99% not, this is probably leftovers due to type inference errors.

File: src/minecraft/block_state.rs (hunk 9 above)
Comment (toqueteos) [inline src/minecraft/block_state.rs]: Reverted.

File: src/main.rs [hunk 15]
```diff
@@ -1,53 +1,54 @@
...
+extern crate quack;
+extern crate render;
 extern crate sdl2;
 extern crate sdl2_window;
-extern crate gfx;
-extern crate device;
-extern crate render;
-#[phase(plugin)]
-extern crate gfx_macros;
-extern crate libc;
```
Comment (eddyb) [inline src/main.rs]: Wheeeee!

**Comment by eddyb** [comment]
Full *steam* ahead! :steam_locomotive:

**Comment by bvssvni** [comment]
Oh yeah!

**Comment by bvssvni** [comment]
Tested on OSX, it works!
